### DISCUSSION TITLE
How bcast works

### DISCUSSION THREAD

**Discussion by hipudding** [ggml-org/llama.cpp#6190] [Q&A] [answered]
I saw a testcase in test-backend-ops to test bcast with ADD, MUL and DIV. As far as I know, bcast only works for different dims, and all dims should be same or 1. But this case cast is adding tensor a(16,10,10,10) with b(32,10,10,10). I read the source code, but can't undersand clearly, could some one tell me how it works?

Does it bcast tensor a(16,10,10,10) to (32,10,10,10), and copy first "0-16" to "17-31"?

**Comment by slaren** [comment, marked as answer]
It repeats the second tensor to match the dimension of the first. Eg. `ggml_add(a, b)` is equivalent to `ggml_add(a, ggml_repeat(b, a))`.
